x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plan/QueryPlan.java

```diff
@@ -66,15 +65,22 @@ public List<Expression> expressions() {
     }
 
     /**
-     * Returns the expressions referenced on this query plan node.
+     * The attributes required to be in the {@link QueryPlan#inputSet()} for this plan to be valid.
+     * Excludes generated references.
+     * <p>
+     * E.g. for {@code EVAL x = 2*some_field, y = 2*x} this includes {@code some_field} but neither {@code x} nor {@code y}.
+     * For {@code ENRICH some_policy ON field WITH some_enrich_field} this includes {@code field} but excludes the generated reference
+     * {@code some_enrich_field}.
      */
     public AttributeSet references() {
         if (lazyReferences == null) {
-            lazyReferences = Expressions.references(expressions());
+            lazyReferences = computeReferences();
         }
         return lazyReferences;
     }
 
+    protected abstract AttributeSet computeReferences();
```

Comment:
I would actually prefer not to give this a default implementation.

As long as we add new commands, I'd prefer each plan node to be forced to explicitly declare what it requires from its children. This is more verbose, but consistent with `output()` - and the fact that `references` _had_ the mentioned default implementation led to many plan nodes having wrong implementations, and we built a bunch of workarounds to account for that.

Case in point: when the `Join` node was first added, its `references` implementation was wrong and needed fixing, but this was not immediately apparent at all.

Let me know how you'd like us to proceed. I'm fine with either way, but wanted to lay out my reasoning first.

Reply:
tl;dr - I've read your arguments however I'd prefer to use a default implementations, similar to inputSet, etc...

references and output are different.
The input of each set is well determined - the output set of its children. The same patter makes sense for references(), expressions() etc.. Not so with the output since each input had its own particularities.

As you point out, the issue was a bug in the default implementation - better to fix that in one place instead of having multiple places that do the same thing and potentially replicate the bug.

At the end of the day, it's an arbitrary choice. The style of the code-base is trustful towards the author, that is non-defensive: (for example collections are not copied for input/returns by default), method calls don't perform null checks for every params, etc...

Hence why I opt towards a default implementation that is useful and guiding - adding a new command requires the author to know what they're doing to begin with so I'd use this mindset, at least at this stage of the project.